src/Clapi/Serialisation.hs

```diff
@@ -139,53 +146,196 @@ cvBuilder (ClFloat x) = return $ floatBE x
 cvBuilder (ClDouble x) = return $ doubleBE x
 cvBuilder (ClString x) = builder x
 cvBuilder (ClList vs) = builder vs
--- cvBuilder (ClList cvs) = aggregate <$> mapM build cvs
---   where
---     build cv = sequence (fromChar $ typeTag cv, cvBuilder cv)
---     aggregate bs = let (bs1, bs2) = unzip bs in mconcat bs1 <> mconcat bs2
-
-cvParser :: Char -> Parser ClapiValue
-cvParser 'F' = return $ ClBool False
-cvParser 'T' = return $ ClBool True
-cvParser 't' = ClTime <$> (parser :: Parser Time)
-cvParser 'e' = ClEnum <$> anyWord8
-cvParser 'u' = ClWord32 <$> anyWord32be
-cvParser 'U' = ClWord64 <$> anyWord64be
-cvParser 'i' = ClInt32 <$> fromIntegral <$> anyWord32be
-cvParser 'I' = ClInt64 <$> fromIntegral <$> anyWord64be
-cvParser 'd' = ClFloat <$> wordToFloat <$> anyWord32be
-cvParser 'D' = ClDouble <$> wordToDouble <$> anyWord64be
-cvParser 's' = ClString <$> (parser :: Parser T.Text)
-cvParser 'l' = ClList <$> (parser :: Parser [ClapiValue])
-
-
-taggedEncode :: (Monoid b, Serialisable b) =>
-    (a -> b) -> (a -> CanFail Builder) -> [a] -> CanFail Builder
-taggedEncode derive build xs = derived <<>> built where
-    built = foldl (<<>>) (return mempty) (map build xs)
-    derived = builder $ foldl (<>) mempty (map derive xs)
-
--- FIXME: Repitition of tag chars :-(
-parseTags :: APT.Parser String
-parseTags = APT.many' $ APT.satisfy (APT.inClass "NFTteuUiIdDsl")
 
-instance Serialisable [ClapiValue] where
-    builder = taggedEncode derive cvBuilder where
-        derive cv = [typeTag cv]
-    parser = do
-        typeTags <- composeParsers parser parseTags
-        sequence $ map cvParser typeTags
+instance Serialisable ClapiValue where
+    builder = tdTotalBuilder cvTaggedData cvBuilder
+    parser = tdTotalParser cvTaggedData cvParser
+      where
+        cvParser :: ClapiTypeEnum -> Parser ClapiValue
+        cvParser ClTTime = ClTime <$> (parser :: Parser Time)
+        cvParser ClTEnum = ClEnum <$> anyWord8
+        cvParser ClTWord32 = ClWord32 <$> anyWord32be
+        cvParser ClTWord64 = ClWord64 <$> anyWord64be
+        cvParser ClTInt32 = ClInt32 <$> fromIntegral <$> anyWord32be
+        cvParser ClTInt64 = ClInt64 <$> fromIntegral <$> anyWord64be
+        cvParser ClTFloat = ClFloat <$> wordToFloat <$> anyWord32be
+        cvParser ClTDouble = ClDouble <$> wordToDouble <$> anyWord64be
+        cvParser ClTString = ClString <$> (parser :: Parser T.Text)
+        cvParser ClTList = ClList <$> (parser :: Parser [ClapiValue])
 
 
 encodeListN :: (Serialisable a) => [a] -> CanFail Builder
-encodeListN = taggedEncode (const (1 :: Sum Word16)) builder
+encodeListN items = (lenBuilder $ length items) <<>> (foldl (<<>>) (return mempty) (map builder items))
```

Comment:
Just a thought, not really for this PR, but we could potentially define a `Monoid` instance for `CanFail` (or perhaps `ExceptT` if we go there, although I don't know if it's possible). That would mean instead of worrying about gluing all the things together you could just `<>`. `Maybe a` is already an instance of `Monoid` if `a` is an instance of `Monoid`, so I think it would just be a case of writing the analagous instance for `CanFail`.